[AMD][rocm-openmp-extras] Update for ROCm 4.3.

estewart08 opened the pull request:
- Added new checksums for 4.3.
- Now using llvm-amdgpu ~openmp in order to use the rocm-device-libs
  build as external project in llvm-amdgpu package. We still need
  to pull device-libs in using resource for the build as some headers
  are not installed.
- Updated symlink creation to now remove an existing link if  present
  to avoid issues on partial reinstalls when debugging.
- Adjusted the flang_warning to be a part of Cmake options instead of
  a filter_file for better compatability.
- The dependency on hsa-rocr-dev created some problems as type was changed
  to the default build/link. This issue was because ROCr uses libelf and
  the build of openmp expects elfutils. When link is specified libelf
  was being found in the path first, causing errors. This was
  introduced with the llvm-amdgpu external project build of device-libs.
- On a bare bone installation of sles15 it was noted that libquadmath0 was
  needed as a dependency. On 18.04 gcc-multilib was also needed.

estewart08 committed:
[AMD][rocm-openmp-extras] Update for ROCm 4.3.

- Added new checksums for 4.3.
- Now using llvm-amdgpu ~openmp in order to use the rocm-device-libs
  build as external project in llvm-amdgpu package. We still need
  to pull device-libs in using resource for the build as some headers
  are not installed.
- Updated symlink creation to now remove an existing link if  present
  to avoid issues on partial reinstalls when debugging.
- Adjusted the flang_warning to be a part of Cmake options instead of
  a filter_file for better compatability.
- The dependency on hsa-rocr-dev created some problems as type was changed
  to the default build/link. This issue was because ROCr uses libelf and
  the build of openmp expects elfutils. When link is specified libelf
  was being found in the path first, causing errors. This was
  introduced with the llvm-amdgpu external project build of device-libs.
- On a bare bone installation of sles15 it was noted that libquadmath0 was
  needed as a dependency. On 18.04 gcc-multilib was also needed.

estewart08 commented:
Any suggestions on adding libquadmath (libquadmath0 for sles15) and gcc-multilib as dependencies? I was unable to find the necessary packages in spack, but I could be overlooking them, thanks.

haampie commented:
Sorry, missed the notification for this one. Generally lgtm

`libquadmath0` is a compiler dependency, and we currently don't model compilers are packages, so the user is responsible for installing that.

Is `libhsa-runtime64.so.1` not being linked to?

Would it be possible to check with your team not to use conflicting elf libs but support either? In that case we use `depends_on('elf')` which is the virtual package and is provided by elfutils and libelf. Btw, is https://github.com/spack/spack/pull/25588#discussion_r695707637 related to this?

estewart08 commented:
> Would it be possible to check with your team not to use conflicting elf libs but support either? In that case we use `depends_on('elf')` which is the virtual package and is provided by elfutils and libelf. Btw, is [#25588 (comment)](https://github.com/spack/spack/pull/25588#discussion_r695707637) related to this?

I can ask around about this. I am unaware of what the issue is for 25588, probably need someone else to respond on that.

> Is `libhsa-runtime64.so.1` not being linked to?
> 
We link in libhsa-runtime64 via cmake for libomptarget.rtl.amdgpu.so in the build of openmp. Adding -L and -I, which is what I understand 'link' does in 'depends on', does not seem to have a benefit, but in this case it creates a build error due to elf issues.

haampie commented:
> Adding -L and -I, which is what I understand 'link' does in 'depends on',

It's not just that, it's that build dependencies can be uninstalled with `spack gc`, so you want to avoid that that removes a lib you actually link to

srekolam commented:
@estewart08  , #25588 while updating the rocm-debug-agent, we found that the later versions (0.168 above) of elfutils is not compatible with rocm-debug-agent. it was giving a build errors. Hence we restricted to use the elfutils@0.168 .

estewart08 committed:
Workaround for libelf headers being used instead of elfutils.

SPACK_INCLUDE_DIRS contains the path for libelf before
elfutils. Until hsa-rocr-dev uses elfutils we can pass
the elfutils include directory via cmake options as a
workaround.

hsa-rocr-dev dependency was reverted to the default
build/link deptype.

estewart08 commented:
Updated the patch with a workaround for the libelf/elfutils issue. The main issue was libelf was found before elfutils in SPACK_INCLUDE_DIRS. Instead of modifying the env variable I chose to pass the elfutils include path via cmake for the build of openmp. This can be limited to certain versions once libelf is removed from hsa-rocr-dev.

estewart08 commented:
@haampie, Can this get a reviewed again? I am wanting to make updates for 4.3.1 and would prefer this land before hand.

srekolam commented:
@haampie ,can we get this  review completed.. this package needs to be updated for 4.3.1 as well.

srekolam commented:
@haampie , @tldahlgren  can I get this reviewed and merged . Been waiting for long.

srekolam commented:
@haampie ..just a reminder.. this recipe requires update for rocm-4.3.1 too
